### PR TITLE
Phase 6: Tandem DSDF Channel Mixup — Synthetic Foil-Shape Interpolation

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The primary bottleneck is **p_tan = 30.53** (surface MAE on NACA6416 tandem transfer). NACA6416 is an unseen foil geometry during training — the training set contains NACA0012 and similar profiles. The model has to generalize its tandem pressure predictions to an out-of-distribution foil shape.

Our data augmentation strategy (gap/stagger aug, DSDF magnitude aug) has proven effective at improving tandem generalization. This experiment extends that principle to **foil-shape space**: by interpolating the DSDF feature channels between two tandem samples, we create synthetic training samples with intermediate foil geometries that span the space between the known training shapes.

The key insight: DSDF channels encode the geometry of each foil (signed distance function gradients = surface normal directions and curvature). A linear interpolation of these channels creates a plausible intermediate geometry that doesn't correspond to any real sample in the training set, but sits between the known training shapes in feature space. NACA6416 sits in the same shape space — this augmentation increases the probability that the model has seen something similar during training.

**Why this won't corrupt the physics:**
- We only mix DSDF channels (x[:,:,2:10]) — the geometric encoding
- Position coordinates, Re/AoA scalars, gap/stagger, and ALL targets are kept fixed
- The physical targets (velocity field, pressure) remain consistent with the base sample's geometry
- This is a well-known technique in computer vision (Cutmix/Mixup) adapted to physics-informed feature spaces

**Expected impact:** -1% to -3% p_tan. Low risk — worst case is the model ignores the augmentation.

## Instructions

In `cfd_tandemfoil/train.py`, add the Tandem DSDF Channel Mixup augmentation.

**1. Add config fields** (near line ~961, after `aug_gap_stagger_sigma`):
```python
aug_tandem_dsdf_mixup: bool = False
aug_tandem_dsdf_mixup_alpha: float = 0.7  # Beta distribution alpha
```

**2. Add augmentation block** in the training loop, immediately after the gap/stagger aug block (~line 1554), BEFORE `x = (x - stats["x_mean"]) / stats["x_std"]`:

```python
# Tandem DSDF Channel Mixup (tandem-only)
if cfg.aug_tandem_dsdf_mixup:
    _is_tan_mixup = (x[:, 0, 22].abs() > 0.01)  # gap feature nonzero -> tandem
    if _is_tan_mixup.sum() >= 2:
        _tan_idx = _is_tan_mixup.nonzero(as_tuple=False).squeeze(1)
        _perm = _tan_idx[torch.randperm(len(_tan_idx), device=x.device)]
        _beta_dist = torch.distributions.Beta(
            torch.tensor(cfg.aug_tandem_dsdf_mixup_alpha, device=x.device),
            torch.tensor(0.4, device=x.device)
        )
        _lam = _beta_dist.sample((_tan_idx.shape[0],)).view(-1, 1, 1)
        # Mix ONLY DSDF channels (indices 2:10) — positions, scalars, targets unchanged
        x[_tan_idx, :, 2:10] = (
            _lam * x[_tan_idx, :, 2:10] +
            (1 - _lam) * x[_perm, :, 2:10]
        )
```

**3. Verify channel indices**: Confirm in `data/prepare_multi.py` that indices 2:10 in raw x are the full DSDF block (foil-1 and foil-2 signed distance channels). The comment at train.py line 1556 (`raw_dsdf = x[:, :, 2:10]`) strongly suggests this is correct.

**Implementation notes:**
- Apply BEFORE standardization — operating in raw feature space
- Apply to ALL tandem samples in the batch (not a subset)
- `_perm` is a random permutation of tandem indices — each tandem sample mixed with a different partner
- Beta(0.7, 0.4): asymmetric, ~70% of samples get lambda > 0.5 (subtle perturbations most of time)
- Training only — no change to eval/validation code
- Single-foil samples (gap=0) completely unaffected

**Experiment sweep — 4 runs total (2 alpha values x 2 seeds):**

```bash
# Alpha=0.7 — seed 42
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/dsdf-mixup-a07-s42" --wandb_group phase6/tandem-dsdf-mixup \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 \
  --aug_tandem_dsdf_mixup --aug_tandem_dsdf_mixup_alpha 0.7

# Alpha=0.7 — seed 43
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/dsdf-mixup-a07-s43" --wandb_group phase6/tandem-dsdf-mixup \
  --seed 43 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 \
  --aug_tandem_dsdf_mixup --aug_tandem_dsdf_mixup_alpha 0.7

# Alpha=0.5 — seed 42
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/dsdf-mixup-a05-s42" --wandb_group phase6/tandem-dsdf-mixup \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 \
  --aug_tandem_dsdf_mixup --aug_tandem_dsdf_mixup_alpha 0.5

# Alpha=0.5 — seed 43
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/dsdf-mixup-a05-s43" --wandb_group phase6/tandem-dsdf-mixup \
  --seed 43 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 \
  --aug_tandem_dsdf_mixup --aug_tandem_dsdf_mixup_alpha 0.5
```

**Decision rule:** Report results for both alpha values. If alpha=0.7 2-seed avg gives p_tan < 30.53, that's a win. If alpha=0.7 regresses p_tan by >0.3 vs baseline after 100 epochs, abort those runs and focus on alpha=0.5 only.

## Baseline

Current 8-seed combined baseline (PR #2123):

| Metric | Baseline (8-seed mean) |
|--------|----------------------|
| p_in   | 13.24 ± 0.33         |
| p_oodc | 7.73 ± 0.22          |
| **p_tan** | **30.53 ± 0.50**  |
| p_re   | 6.50 ± 0.07          |

**Target: p_tan < 30.53 (primary), p_oodc < 7.73 (secondary).**

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results: Tandem DSDF Channel Mixup

### Configuration
- `--aug_tandem_dsdf_mixup` with alpha=0.7 and alpha=0.5
- Seeds 42 and 43 for each alpha, 154-155 epochs reached (180-min timeout)
- W&B group: `phase6/tandem-dsdf-mixup`
- All baseline flags included (`--aft_foil_srf --aug_gap_stagger_sigma 0.02`)

### Results

| Run | alpha | seed | p_in | p_tan | p_oodc | p_re | W&B |
|-----|-------|------|------|-------|--------|------|-----|
| Baseline (8-seed) | — | — | 13.24 | **30.53** | 7.73 | 6.50 | — |
| dsdf-mixup | 0.7 | 42 | **12.4** | 31.6 | **7.6** | **5.5** | [6asz230g](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/6asz230g) |
| dsdf-mixup | 0.7 | 43 | 13.2 | 30.3 | 8.1 | 6.8 | [2f5h1bqy](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/2f5h1bqy) |
| dsdf-mixup | 0.5 | 42 | 13.1 | 31.0 | 7.8 | 6.0 | [oladith7](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/oladith7) |
| dsdf-mixup | 0.5 | 43 | 13.9 | 30.5 | **9.5** | 6.7 | [v5b43gi4](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/v5b43gi4) |

### 2-Seed Averages vs Baseline

| Metric | Baseline | α=0.7 avg | Δ | α=0.5 avg | Δ |
|--------|----------|-----------|---|-----------|---|
| p_in   | 13.24    | 12.80     | **-3.3%** | 13.50 | +2.0% |
| p_tan  | 30.53    | 30.95     | +1.4% | 30.75 | +0.7% |
| p_oodc | 7.73     | 7.85      | +1.6% | 8.65 | **+11.9%** |
| p_re   | 6.50     | 6.15      | **-5.4%** | 6.35 | **-2.3%** |

### Analysis

**Mixed results with high variance.**

**Positive signals:**
- p_re improved significantly — α=0.7 seed 42 achieved **p_re=5.5** (vs 6.50 baseline, -15.4%), and the 2-seed average was -5.4%. The DSDF mixing creates implicit diversity in the geometry-Re interaction space.
- p_in improved for α=0.7 (12.8 avg vs 13.24 baseline, -3.3%).

**Negative signals:**
- p_tan (primary target) **did not improve** — both alpha values show slight regression (+0.7% to +1.4%). The DSDF mixing between tandem samples doesn't help generalize to NACA6416 — it creates interpolations between NACA0012 samples, not new foil shapes.
- p_oodc regressed badly for α=0.5 seed 43 (9.5 vs 7.73, +22.9%). Very high variance between seeds.
- Overall, the augmentation increases training noise without providing consistent directional improvement on the target metric.

**Why p_tan didn't improve:**
The fundamental issue is that mixing DSDFs between two NACA0012 tandem samples creates another NACA0012-like geometry, not an intermediate toward NACA6416. The interpolation stays within the training distribution's hull rather than extending toward OOD shapes. To improve p_tan, augmentation would need to create genuinely novel foil geometries.

### Suggested Follow-ups
- **Larger DSDF perturbation** (random scaling/noise, not mixup): Add multiplicative or additive noise directly to DSDF channels to create genuinely new geometries rather than interpolations
- **Cutmix on spatial patches**: Mix DSDF from different spatial regions to create chimera geometries with mixed pressure/curvature profiles
- **Focus on what works**: The p_re improvement is interesting — DSDF mixing may help specifically for Re generalization through an indirect mechanism (geometry-conditioned Re smoothing)

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2132 because ADVISOR: Closing — p_tan (primary target) regresses in both alpha variants (+1.4% for α=0.7, +0.7% for α=0.5 vs old baseline; +3.5% and +2.8% vs current baseline). High variance (α=0.5/s43 p_oodc=9.47, catastrophic). Student correctly identified root cause: interpolating NACA0012 DSDFs creates more NACA0012-like shapes, not novel geometries toward NACA6416. p_re improvement is interesting but secondary. Good analysis, thorfinn — reassigning to a higher-priority experiment..
